node/cancellation-v2.md

```diff
@@ -0,0 +1,139 @@
+## Issues with the current design
+1. The concept of scopes is an additional term users need to learn which is foreign to JS developers
+2. With the current design there's no way to share or transfer promises between scopes.
+   1. In this example the Workflow cannot be cancelled and will be stuck forever:
+      ```ts
+      // Resolvable promise (should be resolved elsewhere in the code, e.g from signal)
+      let resolve: ((val: any) => void) | undefined = undefined;
+      const p = new Promise((resolve_, reject) => {
+        resolve = resolve_;
+      });
+
+      export async function main(): Promise<void> {
+        await p;
+      }
+      ```
+   1. When trying to cancel the promise below we get an error from WF internals: `Expected to find child scope mapping, got undefined`
+      ```ts
+      import { cancel, cancellationScope, sleep } from '@temporalio/workflow';
+      
+      export async function main(): Promise<void> {
+        const timer = sleep(10);
+        const promise = cancellationScope(async () => {
+          await timer;
+        });
+        cancel(promise);
+        await promise;
+      }
+      ```
+
+## Changes to the current design
+
+### Naming and structure
+
+To solve issue (1), in this proposal the term "scope" goes away and blends in with the concept of Workflow `Context`.
+
+```
+cancellationScope -> Context.cancellable
+shield -> Context.nonCancellable
+cancel -> ContextualPromise.cancel (not related to (1), more on this below)
+```
+
+### New concepts
+
+With these new concepts we can address issue (2).
+
+#### `ContextualPromise` (or `CancellablePromise`)
+
+- Implements the `PromiseLike` interface (awaitable via `then()`), exposes the `cancel` method
+- Context aware: linked to parent context and propagates cancellation, transferable between different contexts
+- Returned by activities, timers, and `Context.cancellable`
+
+#### `ResolvablePromise` (or `Trigger`)
+
+- Extends `ContextualPromise`
+- Provides additional public methods to resolve and reject itself
+- Useful for e.g waiting for a signal from WF main
+
+  ```ts
+  const unblocked = new ResolvablePromise<void>();
+  const signals = {
+    unblock(): void {
+      unblocked.resolve();
+    },
+  };
+
+  async function main() {
+    await unblocked;
+  }
+  ```
+
+### Open Questions
+
+#### Sharing promises between different contexts
+
+1. What happens when a timer or activity is created in a cancellable context and awaited in a non-cancellable context?
+
+```ts
+const p = sleep(3);
+await Context.nonCancellable(async () => {
+  // p is cancellable in a non cancellable context
+  await p;
+});
+```
+
+If the Workflow is cancelled, the user might be surprised if `await p` throws.
+
+2. What happens when a timer or activity was created in a non-cancellable context and awaited in a cancellable context?
+
+```ts
+let p: Promise<void> | undefined = undefined;
+await Context.nonCancellable(async () => {
+  p = sleep(3);
+});
```

Comment:
I'm a little confused about this example here. Is it missing an `await p;` at the end of this closure too, since we're talking about it being awaited in both places?

Reply:
I thought about this a bit more and one of the problems with this design is that promises receive their context on creation instead of when awaited on.
I'm going to change this so promises are only associated with a context once awaited, that should clear up the confusion with this example.

Reply:
This new version makes more sense to me :+1: